core/kernel/console.c

```diff
@@ -54,3 +60,104 @@ void register_serial_console(struct serial_chip *chip)
 {
 	serial_console = chip;
 }
+
+#ifdef CFG_DT
+
+/*
+ * Check if the /secure-chosen node in the DT contains an stdout-path value
+ * for which we have a compatible driver. If so, switch the console to
+ * this device.
+ */
+void configure_console_from_dt(unsigned long phys_fdt)
+{
+	const struct dt_driver *drv, *found = NULL;
+	const struct serial_driver *sdrv;
+	const struct fdt_property *prop;
+	struct serial_chip *dev;
+	char *stdout_data;
+	const char *uart;
+	const char *parms = NULL;
+	void *fdt;
+	int offs;
+	char *p;
+	int st;
+
+	if (!phys_fdt)
+		return;
+	fdt = phys_to_virt(phys_fdt, MEM_AREA_IO_NSEC);
+	if (!fdt)
+		panic();
+
+	offs = fdt_path_offset(fdt, "/secure-chosen");
+	if (offs < 0)
+		return;
+	prop = fdt_get_property(fdt, offs, "stdout-path", NULL);
+	if (!prop) {
+		/*
+		 * /secure-chosen node present but no stdout-path property
+		 * means we don't want any console output
+		 */
+		IMSG("Switching off console");
+		register_serial_console(NULL);
+		return;
+	}
+
+	stdout_data = strdup(prop->data);
+	if (!stdout_data)
+		return;
+	for (p = stdout_data; *p; p++) {
+		if (*p == ':') {
+			*p = '\0';
+			parms = p + 1;
+			break;
+		}
+	}
+
+	/* stdout-path may refer to an alias */
+	uart = fdt_get_alias(fdt, stdout_data);
+	if (!uart) {
+		/* Not an alias, assume we have a node path */
+		uart = stdout_data;
+	}
+	offs = fdt_path_offset(fdt, uart);
+	if (offs < 0)
+		goto out;
+
+	for_each_dt_driver(drv) {
+		const struct dt_device_match *dm;
+
+		for (dm = drv->match_table; dm; dm++) {
+			st = fdt_node_check_compatible(fdt, offs,
+						       dm->compatible);
+			if (!st) {
+				found = drv;
+				break;
+			}
+		}
+	}
+	if (!found)
+		goto out;
+
+	sdrv = (const struct serial_driver *)found->driver;
+	if (!sdrv)
+		goto out;
+	dev = sdrv->dev_alloc();
+	if (!dev)
+		goto out;
+	/*
+	 * If the console is the same as the early console, dev_init() might
+	 * clear pending data. Flush to avoid that.
+	 */
+	console_flush();
+	if (sdrv->dev_init(dev, fdt, offs, parms) < 0) {
+		sdrv->dev_free(dev);
+		goto out;
+	}
+
+	IMSG("Switching console to device: %s", uart);
```

Comment:
Not after the new console has registered, because I want it to appear on the "old" console (assuming they're different) to help the user understand where the subsequent output will go.

About moving it before `console_flush()`: we don't know yet if the init of the new console is OK (and we don't want to print the message if it failed). Would you want a `console_flush()` after this `DMSG()`? Seems pretty useless.